src/adapter/src/coord/read_policy.rs

```diff
@@ -51,27 +38,38 @@ use timely::progress::Antichain;
 
 use crate::coord::id_bundle::CollectionIdBundle;
 use crate::coord::timeline::{TimelineContext, TimelineState};
+use crate::coord::Coordinator;
 use crate::session::Session;
 use crate::util::ResultExt;
 
-/// Relevant information for acquiring or releasing a bundle of read holds.
+/// For each timeline, we hold one [TimelineReadHolds] as the root read holds
+/// for that timeline. Even if there are no other read holds ([ReadHolds] and/or
+/// [ReadHoldsInner]), it acts as a backstop that makes sure that collections
+/// remain readable at the read timestamp (according to the
+/// timestamp oracle) of that timeline.
+///
+/// When creating a collection in a timeline, it is added to the one
+/// [TimelineReadHolds] of that timeline and when a collection is dropped it is
+/// removed.
 ///
-/// See module-level documentation about the split between [InternalReadHolds]
-/// and [ReadHolds].
+/// A [TimelineReadHolds] is never released, it is only dropped when the
+/// corresponding timeline is dropped, which only happens when all collections
+/// in it have been dropped. We only add collections, remove collections, and
+/// downgrade (update, yes!) the read holds.
 #[derive(Debug, Serialize)]
-pub struct InternalReadHolds<T> {
+pub struct TimelineReadHolds<T> {
```

Comment:
This might be a bad idea, but you could turn this into a type alias like `pub type TimelineReadHolds<T> = ReadHoldsInner<T>;` to avoid duplicating all the methods.

Reply:
I'd like to keep them separate: `ReadHoldsInner` has very few methods, and actually only one real `pub` method. And it doesn't have the dangerous `remove_compute_ids` and `remove_storage_ids` that `TimelineReadHolds` has. I say dangerous because those would allow leaking read holds when used incorrectly. :sweat_smile: